internal/worker/authenticationworker/worker.go

```diff
@@ -42,51 +43,65 @@ func NewWorker(st *keyupdater.State, agentConfig agent.Config) (worker.Worker, e
 	if !ok {
 		return nil, errors.NotValidf("machine tag %v", agentConfig.Tag())
 	}
+	keyUpdater := &keyUpdater{
+		st:  st,
+		tag: machineTag,
+	}
+	err := keyUpdater.initializeAuthorizedKeys()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
 	w, err := watcher.NewNotifyWorker(watcher.NotifyConfig{
-		Handler: &keyupdaterWorker{
-			st:  st,
-			tag: machineTag,
-		},
+		Handler: keyUpdater,
 	})
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return w, nil
+
+	return AuthWorker{
+		NotifyWorker: w,
+	}, nil
 }
 
-// SetUp is defined on the worker.NotifyWatchHandler interface.
-func (kw *keyupdaterWorker) SetUp() (watcher.NotifyWatcher, error) {
+// initializeAuthorizedKeys reads the current ssh keys from the state
+// and the current ssh keys from ~/.ssh/authorized_keys. It removes any
+// ephemeral keys.
+func (kw *keyUpdater) initializeAuthorizedKeys() error {
 	// Record the keys Juju knows about.
 	jujuKeys, err := kw.st.AuthorisedKeys(kw.tag)
 	if err != nil {
 		err = errors.Annotatef(err, "reading Juju ssh keys for %q", kw.tag)
 		logger.Infof(err.Error())
-		return nil, err
+		return err
 	}
-	kw.jujuKeys = set.NewStrings(jujuKeys...)
-
+	kw.jujuKeys = set.NewStrings(ensureJujuCommentForKeys(jujuKeys)...)
+	allKeys := kw.jujuKeys.Values()
 	// Read the keys currently in ~/.ssh/authorised_keys.
 	sshKeys, err := ssh.ListKeys(SSHUser, ssh.FullKeys)
 	if err != nil {
 		err = errors.Annotatef(err, "reading ssh authorized keys for %q", kw.tag)
 		logger.Infof(err.Error())
-		return nil, err
+		return err
 	}
 	// Record any keys not added by Juju.
 	for _, key := range sshKeys {
 		_, comment, err := ssh.KeyFingerprint(key)
 		// Also record keys which we cannot parse.
 		if err != nil || !strings.HasPrefix(comment, ssh.JujuCommentPrefix) {
-			kw.nonJujuKeys = append(kw.nonJujuKeys, key)
+			allKeys = append(allKeys, key)
 		}
 	}
```

Comment:
I think we need to expand this out to be a little more verbose. Also handle filtering out the ephemeral keys here.

```go
// Preserve any non-juju keys, including malformed ones.
for _, key := range sshKeys {
	_, comment, err := ssh.KeyFingerprint(key)
	if err != nil {
		logger.Warningf("preserved malformed ssh key %q: %v", key, err)
		allKeys = append(allKeys, key)
		continue
	}
	if strings.HasPrefix(comment, ssh.JujuCommentPrefix) {
		logger.Infof("filtered out juju ssh key %q", key)
		continue
	}
	if strings.HasPrefix(comment, jujuEphemeralCommentPrefix) {
		logger.Infof("filtered out juju ephemeral ssh key %q", key)
		continue
	}
	logger.Infof("preserved ssh key %q", key)
	allKeys = append(allKeys, key)
}
```